File: tests/TestBindings/TestFilters.cpp
#include <catch.hpp>
#include <sol.hpp>
#include <BWAPI_Lua.h>
#include "BindingTestHelpers.h"
#include "MockUnit.h"

using namespace BWAPI_Lua;

TEST_CASE("Filters", "[filter]")
{
	sol::state lua;
	lua.open_libraries(sol::lib::base);
	auto BWAPI = lua.create_named_table("BWAPI");
	bindUnit(BWAPI);
	TestBindings::bindMockUnitImpl(BWAPI);
	bindFilters(BWAPI);
	lua.script_file(TestBindings::getLuaTestPath(__FILE__));

	SECTION("Lowest")
	{
		sol::function test = lua["lowest"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("Highest")
	{
		sol::function test = lua["highest"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("Unary unit filters")
	{
		sol::function test = lua["unaryUnitFilter"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("Comparison unit filters")
	{
		sol::function test = lua["comparisonUnitFilter"];
		REQUIRE_NOTHROW(test());
	}
}

File: tests/TestBindings/Lua/TestFilters.lua
unit1 = BWAPI.MockUnit.new()
unit2 = BWAPI.MockUnit.new()

objs = {
  {val=5}, {val=4}, {val=3}, {val=10}, {val=1}, {val=7}
}
local objValFilter = function(obj) return obj.val end

function lowest()
  local bestFilter = BWAPI.Lowest(objValFilter)
  local best = nil

  for _, obj in ipairs(objs) do
    if best == nil then
      best = obj
    else
      best = bestFilter(best, obj)
    end
  end

  assert(best ~= nil)
  assert(best.val == 1)
end

function highest()
  local bestFilter = BWAPI.Highest(objValFilter)
  local best = nil

  for _, obj in ipairs(objs) do
    if best == nil then
      best = obj
    else
      best = bestFilter(best, obj)
    end
  end

  assert(best ~= nil)
  assert(best.val == 10)
end

function unaryUnitFilter()
  local isTransport = BWAPI.Filter.IsTransport(unit1)
  assert(isTransport == false)
end

function comparisonUnitFilter()
  -- int
  local hp = BWAPI.Filter.HP(unit1)
  assert(hp == 0)

  -- WeaponType
  local weaponType = BWAPI.Filter.GroundWeapon(unit1)
  assert(type(weaponType) == 'userdata')
end
